x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/conditional/SpatialRelatesFunction.java

```diff
@@ -0,0 +1,499 @@
+        ) {
+            super(factoryCreator);
+        }
+
+        @Override
+        public EvalOperator.ExpressionEvaluator.Factory get(Function<Expression, EvalOperator.ExpressionEvaluator.Factory> toEvaluator) {
+            try {
+                return factoryCreator.apply(
+                    source(),
+                    asGeometryDocValueReader(spatialDataType, left()),
+                    asLuceneComponent2D(spatialDataType, right())
+                );
+            } catch (IOException e) {
+                throw new IllegalArgumentException("Failed to process spatial constant", e);
+            }
+        }
+    }
+
+    protected record SpatialEvaluatorKey(DataType dataType, boolean useDocValues, boolean leftIsConstant, boolean rightIsConstant) {
+        UnsupportedOperationException unsupported() {
+            return new UnsupportedOperationException("Unsupported spatial relation combination: " + this);
+        }
+    }
+
+    protected static class SpatialRelations {
+        protected final ShapeField.QueryRelation queryRelation;
+        protected final SpatialCoordinateTypes spatialCoordinateType;
+        protected final CoordinateEncoder coordinateEncoder;
+        protected final ShapeIndexer shapeIndexer;
+        protected final DataType spatialDataType;
+
+        protected SpatialRelations(
+            ShapeField.QueryRelation queryRelation,
+            SpatialCoordinateTypes spatialCoordinateType,
+            CoordinateEncoder encoder,
+            ShapeIndexer shapeIndexer
+        ) {
+            this.queryRelation = queryRelation;
+            this.spatialCoordinateType = spatialCoordinateType;
+            this.coordinateEncoder = encoder;
+            this.shapeIndexer = shapeIndexer;
+            this.spatialDataType = spatialCoordinateType.equals(SpatialCoordinateTypes.GEO) ? GEO_SHAPE : CARTESIAN_SHAPE;
+        }
+
+        protected boolean geometryRelatesGeometry(BytesRef left, BytesRef right) throws IOException {
+            Geometry rightGeom = fromBytesRef(right);
+            if (rightGeom instanceof Point point) {
+                Geometry leftGeom = fromBytesRef(left);
+                return pointRelatesGeometry(point, leftGeom);
+            } else {
+                // Convert one of the geometries to a doc-values byte array, and then visit the other geometry as a Component2D
+                Component2D rightComponent2D = makeComponent2D(rightGeom);
+                return geometryRelatesGeometry(left, rightComponent2D);
+            }
+        }
+
+        protected Geometry fromBytesRef(BytesRef bytesRef) {
+            try {
+                return SpatialCoordinateTypes.UNSPECIFIED.wkbToGeometry(bytesRef);
+            } catch (IllegalArgumentException e) {
+                // TODO: would be better to not rely on exceptions to tell the difference between WKB and WKT
+                return SpatialCoordinateTypes.UNSPECIFIED.wktToGeometry(bytesRef.utf8ToString());
```

Comment:
This is not good. Weshould know at every moment if we are dealing with WKT or WKB, otherwise we are loosing control of the situation.

Reply:
The function should have enough type information to know this. I agree, we should know and call the right one every time.

FWIW at evaluation time we don't annotate the `BytesRefBlock`s with their field type - but we still know the type because that's all set up at planning time.